### PR TITLE
[BugFix][Enhancement] skip building ht in hash join build side if query is cancelled

### PR DESCRIPTION
**Pull request by silverbullet233**
## What type of PR is this:

- [x] BugFix
- [ ] Feature
- [x] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

Does this PR entail a change in behavior?

- [ ] Yes, this PR will result in a change in behavior.
- [x] No, this PR will not result in a change in behavior.

If yes, please specify the type of change:

- [ ] Interface/UI changes: syntax, type conversion, expression evaluation, display information
- [ ] Parameter changes: default values, similar parameters but with different default values
- [ ] Policy changes: use new policy to replace old one, functionality automatically enabled
- [ ] Feature removed
- [ ] Miscellaneous: upgrade & downgrade compatibility, etc.

## Checklist:

- [ ] I have added test cases for my bug fix or my new feature
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function

## Bugfix cherry-pick branch check:

- [x] I have checked the version labels which the pr will be auto-backported to the target branch
  - [x] 3.2
  - [x] 3.1
  - [x] 3.0
  - [ ] 2.5

### CODE REVIEW

File: be/src/exec/hash_join_components.cpp [hunk 1]
```diff
@@ -87,8 +87,11 @@ Status HashJoinBuilder::append_chunk(RuntimeState* state, const ChunkPtr& chunk)
...
 
 Status HashJoinBuilder::build(RuntimeState* state) {
-    SCOPED_TIMER(_hash_joiner.build_metrics().build_ht_timer);
-    TRY_CATCH_BAD_ALLOC(RETURN_IF_ERROR(_ht.build(state)));
+    // if query is already cancelled, building hash table is unnecessary, just skip it
+    if (!state->is_cancelled()) {
+        SCOPED_TIMER(_hash_joiner.build_metrics().build_ht_timer);
+        TRY_CATCH_BAD_ALLOC(RETURN_IF_ERROR(_ht.build(state)));
+    }
     _ready = true;
```
Comment (kangkaisen) [inline be/src/exec/hash_join_components.cpp:95]: if state->is_cancelled, _ready should be false?

**Review by satanson — DISMISSED** [review]

**Comment by wanpengfei-git** [comment]
#### [FE Incremental Coverage Report]


:heart_eyes: **pass** : 0 / 0 (0%)

File: be/src/exec/hash_join_components.cpp [hunk 2]
```diff
@@ -87,8 +87,11 @@ Status HashJoinBuilder::append_chunk(RuntimeState* state, const ChunkPtr& chunk)
...
     return Status::OK();
```
Comment (fzhedu) [inline be/src/exec/hash_join_components.cpp:96]: still return ok if cancelled?

File: be/src/exec/hash_join_components.cpp (hunk 1 above)
Comment (silverbullet233) [reply be/src/exec/hash_join_components.cpp:95]: > if state->is_cancelled, _ready should be false?

_ready was introduced when implementing spill, but I didn't see it being called anywhere. It should have no impact. I'll confirm it with @stdpain later.

File: be/src/exec/hash_join_components.cpp (hunk 2 above)
Comment (silverbullet233) [reply be/src/exec/hash_join_components.cpp:96]: > still return ok if cancelled?

yeah, my idea is to skip this process only, leaving an empty hash table here, and keep other processes the same as before, so the risk is relatively small.

Comment (silverbullet233) [reply be/src/exec/hash_join_components.cpp:96]: > > still return ok if cancelled?
> 
> yeah, my idea is to skip this process only, leaving an empty hash table here, and keep other processes the same as before, so the risk introduced by my change is relatively small.

File: be/src/exec/hash_join_components.cpp (hunk 1 above)
Comment (silverbullet233) [reply be/src/exec/hash_join_components.cpp:95]: > > if state->is_cancelled, _ready should be false?
> 
> _ready was introduced when implementing spill, but I didn't see it being called anywhere. It should have no impact. I'll confirm it with @stdpain later.

confirmed, it is not used for now.

**Comment by wanpengfei-git** [comment]
#### [FE Incremental Coverage Report]


:heart_eyes: **pass** : 0 / 0 (0%)

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]

:disappointed: **fail** : 0 / 2 (00.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|src/exec/pipeline/hashjoin/hash_join_build_operator.cpp|0|2|00.00%|[80, 81]|

**Review by kangkaisen — APPROVED** [review]

**Review by satanson — APPROVED** [review]

**Comment by wanpengfei-git** [comment]
@Mergifyio backport branch-3.2

**Comment by wanpengfei-git** [comment]
@Mergifyio backport branch-3.1

**Comment by wanpengfei-git** [comment]
@Mergifyio backport branch-3.0

**Comment by silverbullet233** [comment]
@mergify backport branch-2.5

**Pull request merged**